File: Basics/1.encapsulation.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    Encapsulation: wrapping up of data and methods into a single unit called class.
    - Data Hiding: hiding the data from outside the class.
    - Access Specifiers: public, private, protected.
*/
class Student
{
    //  Attributes with Access Specifiers
public: // public members are accessible from outside the class
    string name, dept, user_name;
    int age;

private: // private members are not accessible from outside the class
    string password;

protected: // protected members are accessible from derived classes
    string address;

// constructor
public:
    Student(string name, string dept, string user_name, int age, string password, string address)
    {

        /*
            - this pointer is used to refer to the current object
            - this-> means (*this).attributes; dereference and access the attributes
        */
        this->name = name; 
        this->dept = dept;
        this->user_name = user_name;
        this->age = age;
        this->password = password;
        this->address = address;
    }

    // copy constructor: used to copy the attributes of one object to another object
    Student(Student &s)
    {
        name = s.name;
        dept = s.dept;
        user_name = s.user_name;
        age = s.age;
        password = s.password;
        address = s.address;
    }

    //  Member Functions
public:
    void set_password(string pass) // Setter Function
    {
        password = pass;
    }

    string get_password() // Getter Function
    {
        return password;
    }

    void get_info() {
        cout << "Name: " << name << endl;
        cout << "Department: " << dept << endl;
        cout << "User Name: " << user_name << endl;
        cout << "Age: " << age << endl;
        cout << "Address: " << address << endl;
    }
};

int main()
{
    Student s1("John Doe", "CSE", "john_doe", 20, "1234", "Dhaka");
    s1.get_info();
    return 0;
}

File: 1.Basics/1.encapsulation.java
import java.util.*;

/*
    Encapsulation: wrapping up of data and methods into a single unit called class.
    - Data Hiding: hiding the data from outside the class.
    - Access Specifiers: public, private, protected.
*/
class Student {
    //  Attributes with Access Specifiers
    public String name, dept, user_name;
    public int age; // public: can be accessed from anywhere
    private String password; // private: can only be accessed from within the class
    protected String address; // protected: can be accessed from within the class and its subclasses

    // constructor
    public Student(String name, String dept, String user_name, int age, String password, String address) {
        /*
            - this keyword is used to refer to the current object
            - this. means this object's attributes
        */
        this.name = name;
        this.dept = dept;
        this.user_name = user_name;
        this.age = age;
        this.password = password;
        this.address = address;
    }

    // copy constructor: used to copy the attributes of one object to another object
    public Student(Student s) {
        this.name = s.name;
        this.dept = s.dept;
        this.user_name = s.user_name;
        this.age = s.age;
        this.password = s.password;
        this.address = s.address;
    }

    //  Member Functions
    public void setPassword(String pass) { // Setter Function
        this.password = pass;
    }

    public String getPassword() { // Getter Function
        return this.password;
    }

    public void getInfo() {
        System.out.println("Name: " + this.name);
        System.out.println("Department: " + this.dept);
        System.out.println("User Name: " + this.user_name);
        System.out.println("Age: " + this.age);
        System.out.println("Address: " + this.address);
    }
}

public class Encapsulation {
    public static void main(String[] args) {
        Student s1 = new Student("John Doe", "CSE", "john_doe", 20, "1234", "Dhaka");
        s1.getInfo();
    }
}
